### ISSUE TITLE
Can't return damaged item from advanced recipe

### ISSUE THREAD

**Issue by LachimHeigrim** [stanhebben/MineTweaker3#329]
Enviromine mod allows attaching camel pack to armor by crafting the two together.  Info about water remaining in camel pack is stored in armor's NBT tag. Once empty, user can detach the camel pack by crafting a combined armor. Camel pack is returned from crafting, camel pack-less armor stays in crafting grid. 

Enviromine does not correctly detect Natura's armor, so I was trying to add support with Minetweaker instead.

The part that I was not able to implement was to "detach" the camel pack from armor. I.e. if armor has NBT tag of camel pack, I want to return camel pack (with correct water amount remaining) and armor (with correct damage value, if any). 

See the code below. I tried to transform the crafting ingredients in input array first, but it did not work. Then I tried the recipe function approach, with the same result. 

```
val chestplateMud = <BiomesOPlenty:chestplateMud:*>;
val camelPack = <enviromine:camelPack>;

# Works: armor + camel pack  -> armor with camel pack NBT
recipes.addShapeless(
  chestplateMud,
  [chestplateMud.marked("armor"),camelPack.marked("pack")],
  function(output, inputs, crafting) {
    val packTag = inputs.pack.tag;
    return inputs.armor.withTag(packTag);
  }
);

# Broken: armor with camel pack -> camel pack (result) + armor (side-effect) 
# Actual result: Only camel pack is returned. Armor is consumed in the crafting process. 
recipes.addShapeless(
  camelPack,
  [chestplateMud.marked("armor").onlyDamaged().onlyWithTag({camelPath: "enviromine:camelPack"})],
  function(output,inputs,crafting) {
    val packTag = inputs.armor.tag;
    val armorDamage = inputs.armor.damage;
    val giveBackArmor = <BiomesOPlenty:chestplateMud>.withDamage(armorDamage); 
    inputs.armor.giveBack(giveBackArmor);
    return <enviromine:camelPack>.withTag(packTag); 
  }
);
```

**Comment by Voidi** [comment]
I'm not sure if giveBack works **inside** a Recipe function.
I think there is no easy solution for this yet. 
Also this recipes may corrupta existing NBT Data on the Armor.

**Comment by LachimHeigrim** [comment]
I initially tried to use giveBack() as well as transformReplace() on crafting ingredients, i.e. outside of the Recipe function. Did not work either.

What exactly do you mean by damaging the NBT data on Armor? That I need to additionally copy the NBT tag on the armor used as ingredient to preserve it's information?

**Comment by LachimHeigrim** [comment]
I stand corrected. Actually giveBack() on crafting ingredients worked. The problem was, I needed to remove the NBT tag before returning the armor. Otherwise after user picked up the camel back from crafting, the armor stayed in the crafting grid (with camel pack NBT that I could not remove) so user was offered another camel pack. This led to duplication of camel packs.

**Comment by Voidi** [comment]
Right, thats what i want to say. For example if you have an Armor with Enchantments these would be overwritten.
~~What Minetweaker is actually missing are funktions to combine to NBT or remove specific Parts of NBT.~~

Actually there is something like that: updateTag
https://github.com/stanhebben/MineTweaker3/blob/master/MineTweaker3-API/src/main/java/minetweaker/api/item/IItemStack.java#L215

There seems no function to delete a specific key in the NBT, but you could try to set Camelpack key to an empty Array or something similar:
`updateTag({camelPath: ""});` _or_ `updateTag({camelPath: []});`
Depends on how enviromine checks the NBT on the camelpacks.

**Comment by VariableVixen** [comment]
So, we need a removeTag method then?

**Comment by jaredlll08** [comment]
Who wants to test a version of the `"<minecraft:diamond>.removeTag(String name)"`?
if so, what game version?

**Comment by LachimHeigrim** [comment]
I'm testing on 1.7.10. Any chance for test build so that I don't need to build from sources myself?

Would be good for the proposed method to be fluid or vararg as one might need to remove multiple keys. Not sure what are ZenScript's compiler capabilities.

This is `/mt hand` output for enchanted iron chest plate combined with Enviromine camel pack. Being a responsible programmer I'd want to wipe out all keys related to camel pack.

`<minecraft:iron_chestplate>.withTag({ ench: [{lvl: 4 as short, id: 3 as short}], camelPackFill: 24,
 RepairCost: 2, camelPackMax: 100, packName: "enviromine:camelPack"})`

**Comment by jaredlll08** [comment]
ok so, here is the versions:
http://www.blamejared.com/content/CraftTweaker-Customized-1.7.10-3.0.2.jar
http://www.blamejared.com/content/CraftTweaker-Customized-1.8.8-3.0.2.jar
try
`<mod:item>.removeTag("name");`
for example
`minecraft:iron_chestplate.removeTag("ench");

**Comment by LachimHeigrim** [comment]
Sorry I'm not able to download. Ping gets through, but HTTP gets blocked, download is not starting. Looks firewalled.

**Comment by Blue64** [comment]
you could always modify the "max" and "current" to zero, thus it appears to dupe, but the contents are emptied, but upon further review it is revealed that it's essentially removed since you can't actually input any water to the Armor (since the capacity is zero)

**Comment by Blue64** [comment]
another problem with the script that I'm seeing up above is that you are using `val` inside your `loop`, if I'm not mistaken, this would cause problems once it is used in a different fashion, so you will actually desire a `var`.

**Comment by Blue64** [comment]
another potential problem with the script:
```
  [chestplateMud.marked("armor").onlyDamaged().onlyWithTag({camelPath: "enviromine:camelPack"})],
```
`onlyWithTag({camelPath:` camelPa**th**? isn't it supposed to be camelPa**ck**?

**Comment by Blue64** [comment]
also, you'd need to add a recipe for undamaged armor as well, since your input is currently `.onlyDamaged()` which means that if they make a fresh chestplate and then Camel Pack it, that they will be unable to remove it until they damage it.
